app/src/main/java/com/kamron/pogoiv/MainActivity.java

```diff
@@ -145,12 +147,12 @@ protected void onCreate(Bundle savedInstanceState) {
         npTrainerLevel.setWrapSelectorWheel(false);
         npTrainerLevel.setValue(trainerLevel);
 
-        Button launch = (Button) findViewById(R.id.start);
-        launch.setOnClickListener(new View.OnClickListener() {
+        launchButton = (Button) findViewById(R.id.start);
+        launchButton.setOnClickListener(new View.OnClickListener() {
 
             @Override
             public void onClick(View v) {
-                if (((Button) v).getText().toString().equals(getString(R.string.main_permission))) {
+                if (!hasAllPermissions()) {
```

Comment:
Oh, I understand now! I also loved to write in C, but since I started to write Android code I "forgot" that rule.
If you want I can move hasAllPermissions up before onCreate, and I'd do the same with isPokeFlyRunning.
But then I should do the same with startPokeFly and startScreenService... :/

@sarav @Blaisorblade what should I do?